server/routes/api/protein.py

```diff
@@ -25,10 +30,293 @@
+        'id': node_id,
+        'name': protein,
+        'species': species,
+        'depth': depth,
+    }
+
+
+def extract_intactmi(score_dcids):
+    '''
+    Takes list of interaction score DCIDs and numerically parses and returns the first instance
+    of an IntactMi score, or the default interaction score if none are found.
+    
+    TODO: add example
+    '''
+    for score_dcid in score_dcids:
+        if score_dcid.startswith("IntactMiScore"):
+            return float(score_dcid.replace("IntactMiScore", ""))
+    return DEFAULT_INTERACTION_SCORE
+
+
+def interactors(interaction_id_or_dcid):
+    '''
+    Parse and return ids of two interactors from interaction DCID of the form 'bio/{protein1 id}_{protein2 id}'
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' --> 'P53_HUMAN', 'CBP_HUMAN'
+    '''
+    interaction_id = id(interaction_id_or_dcid)
+    protein1, species1, protein2, species2 = interaction_id.split('_')
+    return f'{protein1}_{species1}', f'{protein2}_{species2}'
+
+
+def target(interaction_id_or_dcid, source_id_or_dcid):
+    '''
+    Given interaction DCID and the id one of the interactors, return the other one.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN', 'P53_HUMAN' --> 'CBP_HUMAN'
+    '''
+    interaction_id, source_id = id(interaction_id_or_dcid), id(
+        source_id_or_dcid)
+    interactor_id1, interactor_id2 = interactors(interaction_id)
+    if interactor_id1 == source_id:
+        return interactor_id2
+    return interactor_id1
+
+
+def reverse_interaction_id(interaction_id):
+    '''
+    Given interaction ID of the form '{protein1 id}_{protein2 id}', return '{protein2 id}_{protein1 id}'.
+    E.g. 'P53_HUMAN_CBP_HUMAN' --> 'CBP_HUMAN_P53_HUMAN'
+    '''
+    interactor_id1, interactor_id2 = interactors(interaction_id)
+    return f'{interactor_id2}_{interactor_id1}'
+
+
+def reverse_interaction_dcid(interaction_dcid):
+    '''
+    Given interaction DCID of the form 'bio/{protein1 id}_{protein2 id}', return 'bio/{protein2 id}_{protein1 id}'.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' --> 'bio/CBP_HUMAN_P53_HUMAN'
+    '''
+    return dcid(reverse_interaction_id(id(interaction_dcid)))
+
+
+def filter_interaction_dcids(interaction_dcids):
+    '''
+    Given list of interaction DCIDs, remove self-interactions and deduplicate (keep one copy of each set of duplicates).
+    Self-interactions are interactions DCIDs of the form 'bio/{protein id}_{protein id}' (e.g. 'bio/P53_HUMAN_P53_HUMAN')
+    Duplicates include both exact duplicates and reverse duplicates.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' is duplicate with both itself and 'bio/CBP_HUMAN_P53_HUMAN'.
+    '''
+    dcids = set()
+    uniques = []
+    for interaction_dcid in interaction_dcids:
+        source_id, target_id = interactors(interaction_dcid)
+        # filter out self-interactions
+        if source_id == target_id:
+            continue
+        reverse_dcid = reverse_interaction_dcid(interaction_dcid)
+        if interaction_dcid in dcids or reverse_dcid in dcids:
```

Comment:
So the first time you see an interaction_dcid it won't be in the dcids set, right? And at this point, you won't insert it in uniques. So if you never see that again, you won't return it as part of uniques?

Reply:
Ahh thank you so much for catching this! Yes I meant `interaction_dcids`, but this was very hard to catch because almost everything is repeated in the forward/reverse directions except for self-interactions which we already filtered out.

Reply:
Actually I have no idea what I was thinking! Rewrote the function hopefully it makes more sense now

Reply:
Great!

Reply:
Ideally, we'd have unit tests for all functions in this file. But this can be a TODO if you have some time